### ISSUE TITLE
Function arguments evaluated lazily

### ISSUE THREAD

**Issue by mrmonday** [idris-lang/Idris-dev#3778]
In the following code:

```idris
bottom : ty
bottom = bottom

-- func1 id bottom = 4
func1 : (Int -> Int) -> Int -> Int
func1 f x = 4

-- func2 bottom = bottom
func2 : Int -> Int
func2 x = 4

main : IO ()
main =
  do
    --let x = func1 id bottom
    let x = func2 bottom
    putStrLn (cast x)
```

When running at the REPL, I get:
```
*foo> func1 id bottom
4 : Int
*foo> func2 bottom
^CInterrupt
```
That is, `func1` evaluates its arguments lazily, and `func2` evaluates its arguments strictly.

When compiling with:
```
$ idris foo.idr -o foo
```
Both variants appear to evaluate their arguments lazily:
```
$ ./foo 
4
```

So there are (possibly) two issues here:
 1. The REPL and compiled code are inconsistent
 2. I was under the impression that Idris code is evaluated strictly? If so, all of these scenarios should enter infinite loops.

```
$ idris --version
1.0
```

**Comment by ahmadsalim** [comment]
Thanks for reporting the issue.

**Comment by ahmadsalim** [comment]
I think you are right that one should expect the `func2` call to go into an infinite loop.

**Comment by edwinb** [comment]
It's because the REPL uses the compile time notion of evaluation, which works under the assumption that code is total so can happily make any choice of evaluation order it likes. In practice, this is lazy because the evaluator is using Haskell's evaluation model. 

Nevertheless, it is still useful for test purposes to evaluate non-total things at the REPL, so it does allow that, leading to the behaviour you see here.

In fact, the REPL and compiled code are *necessarily* inconsistent, in that they are different phases of the compiler and the notion of evaluation is therefore slightly different. Compiled Idris code is strict (though even then there may be some optimisations resulting from partial evaluation) but that doesn't necessarily mean that the evaluator at compile time has to be strict.

There is a partial explanation in the FAQ: http://docs.idris-lang.org/en/latest/faq/faq.html#evaluation-at-the-repl-doesn-t-behave-as-i-expect-what-s-going-on

In short, though, this is expected behaviour - if you want to see how compiled code behaves at the REPL, rather than using the compile time notion of evaluation, you need to use `:exec`.

It might be better if the evaluator at the REPL used exactly the same notion of evaluation as the type checker, but I think there are too many situations where this would be irritating or surprising at the moment (e.g. the export rules would mean things not getting reduced).

**Comment by asandroq** [comment]
So, can this be closed?

**Comment by ahmadsalim** [comment]
I will close it as won't fix.
